### PR TITLE
feat(redshift): rewrite Redshift table DDL keywords for a PostgreSQL backend

### PR DESCRIPTION
**Pull request by vuthehuyht**
﻿## Summary

Part 2 of splitting #2836 (`feat/redshift-sql-interceptor`) into smaller, independently reviewable PRs.

Adds `RedshiftSqlInterceptor`: a pure, literal-aware function that rewrites Redshift-specific table DDL clauses (`DISTSTYLE`, `DISTKEY`, `SORTKEY`, `ENCODE`) for compatibility with a plain PostgreSQL backend.

Does not wire the interceptor into the proxy loop yet (deferred to Part 3).

## Type of change

- [x] New feature (`feat:`)

## AWS Compatibility

Rewrites Redshift-specific DDL syntax against the official Redshift `CREATE TABLE` / `ALTER TABLE` grammar (https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html):
- Strips `DISTSTYLE ALL|EVEN|KEY|AUTO`
- Strips `DISTKEY (<col>)`, `DISTKEY <col>`, and column-level `DISTKEY` (including when followed by column constraints `NOT NULL`, `PRIMARY KEY`, `UNIQUE`, `REFERENCES`, etc. or column attributes `ENCODE`)
- Strips `[COMPOUND|INTERLEAVED] SORTKEY (<cols>)`, `SORTKEY <col>`, and column-level `SORTKEY` (including before column constraints)
- Strips `ENCODE <codec>` for valid Redshift column encodings and `ENCODE AUTO`
- Cleans orphaned or double commas from stripped clauses
- Safe scoping and fail-open: only rewrites `CREATE TABLE` and `ALTER TABLE` statements, masks single-quoted and dollar-quoted literals (`$$...$$`, `$tag$...$tag$`) before evaluation, and preserves columns legitimately named `distkey`, `sortkey`, or `encode`

## Checklist

- [x] `./mvnw test` passes locally
- [x] New or updated integration test added
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)

## Context

Part of the Redshift proxy split:
1. [x] **RDS proxy primitives** (#2913, merged)
2. **This PR** - `RedshiftSqlInterceptor`: strip `DISTSTYLE` / `DISTKEY` / `SORTKEY` / `ENCODE` from Redshift table DDL, fail-open.
3. **Wire decoder + intercepting bridge** - `PostgresWireDecoder` and the framed `RedshiftInterceptingBridge` (DDL path only), wired into `RedshiftAuthProxy` / `RedshiftProxyManager`.
4. **S3 COPY simulator** - `CopyStatementParser` + `S3CopySimulator.runCopyFrom`.
5. **S3 UNLOAD simulator** - `S3CopySimulator.runUnload`.

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/redshift/proxy/RedshiftSqlInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+        List<String> literals = new ArrayList<>();
+        String masked = maskStringLiterals(sql, literals);
+
+        String s = DISTSTYLE_PATTERN.matcher(masked).replaceAll("");
+        s = DISTKEY_PAREN_PATTERN.matcher(s).replaceAll("");
+        s = SORTKEY_PAREN_PATTERN.matcher(s).replaceAll("");
+        s = TABLE_LEVEL_KEY_IDENT_PATTERN.matcher(s).replaceAll("");
+        s = KEY_COLUMN_CONSTRAINT_PATTERN.matcher(s).replaceAll("");
+        s = ENCODE_PATTERN.matcher(s).replaceAll("");
```
[the remark this answers is not in the document]
Comment (vuthehuyht) [reply src/main/java/io/github/hectorvent/floci/services/redshift/proxy/RedshiftSqlInterceptor.java]: Fixed. Statements in a multi-query batch are now split on semicolon boundaries (outside string literals) and rewritten independently so non-DDL statements are untouched.

File: docs/services/redshift.md [hunk 2]
```diff
@@ -135,10 +135,25 @@ cluster = redshift.create_cluster(
 print(cluster["Cluster"]["Endpoint"])
 ```
 
+## SQL Interceptor
+
+Floci's Redshift auth proxy intercepts frontend queries on the PostgreSQL wire protocol (Simple Query `'Q'` protocol) to emulate common Redshift-specific SQL syntax:
+
+### Supported Features
+
+- **DDL Compatibility:**
```
[the remark this answers is not in the document]
Comment (vuthehuyht) [reply docs/services/redshift.md]: Reverted docs changes for this PR; documentation will land in Part 3 alongside the proxy wire bridge.

**Review by pgermosen — COMMENTED** [review]
The regex engineering here is genuinely careful, string-literal masking, dollar-quoting, the identity fast path, all stress-tested rather than just written to pass the obvious cases. But there's a real gap: checked `KEY_COLUMN_CONSTRAINT_PATTERN` against the actual [CREATE TABLE grammar](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html), and column_attributes (DISTKEY, SORTKEY, ...) are ordered before column_constraints (NOT NULL, UNIQUE, PRIMARY KEY, REFERENCES) in the official syntax. So `id INT DISTKEY NOT NULL` is valid, ordinary Redshift DDL, but the lookahead here only accepts `,`, `)`, or ENCODE right after the keyword, so DISTKEY is left in place and the rewritten statement still fails on Postgres. A NOT-NULL distribution/sort key column isn't an edge case, and nothing in the test suite exercises it. Worth widening that lookahead (and adding a test) before this becomes the base the rest of the split builds on.

Smaller thing: the body doesn't follow the repo's PR template, and it says this updates `docs/services/redshift.md`, but the diff only touches the interceptor and its test, no docs file made it in.

**Comment by vuthehuyht** [comment]
Good catch, thank you! Widened `KEY_COLUMN_CONSTRAINT_PATTERN`'s lookahead to cover column constraints (`NOT NULL`, `NULL`, `UNIQUE`, `PRIMARY KEY`, `REFERENCES`, `CHECK`, `CONSTRAINT`, etc.) and column attributes following Redshift's `CREATE TABLE` grammar. Also tightened `TABLE_LEVEL_KEY_IDENT_PATTERN` and reordered the passes so column types with precision/length parentheses (e.g. `VARCHAR(10)`) followed by constraints aren't mistakenly matched as table-level key clauses. Added tests covering these variants.
Also aligned the PR body with the repo template and removed the leftover docs reference.

**Review by pgermosen — APPROVED** [review]
Verified this by tracing the regex directly rather than taking "fixed" at face value. The widened lookahead covers NOT NULL and the rest correctly, and the reordering you did was actually necessary for a second reason beyond the one I flagged: traced it by hand-running both orderings against `code VARCHAR(10) DISTKEY UNIQUE`, under the old order, VARCHAR(10)'s own closing paren fools the table-level pattern into misreading `DISTKEY UNIQUE` as a table-level clause and silently drops UNIQUE entirely. The new ordering plus the tightened trailing lookahead closes that too, confirmed by running both versions against that input. Good catch on your end noticing that interaction.

Nothing else outstanding. Good fix.
